[MRG+2]: Copy pilutils functions that are deprecated in scipy

jotasi opened the pull request:
#### Reference Issues/PRs
Fixes  #10147
See also  #10149 which is an incomplete and inactive pull request that attempts to fix the same issue.


#### What does this implement/fix? Explain your changes.
`scipy.misc.pilutil`'s functions are deprecated since version 1.0.0 and will be removed in 1.2.0. As decided in the discussion of #10147 this PR copies the functionality needed in scikit-learn (`imread` and `imresize`) and their respective test cases. Following the discussion in #10149 they are put into a utility module in `sklearn.util.pilutil`. I only included functionality that is already used in scikit-learn, dropping unneeded functions and parameters.


#### Other comments
I'm still unsure about the following points:

1. Licensing: As described in [SciPy's docs](https://github.com/scipy/scipy/blob/master/LICENSE.txt) I copied their license's text into the files containing the implementation as well as the tests. Is this sufficient? I'm especially unsure about the images used in the tests which cannot simply include the license text.

2. Testing: The added tests are not run on Travis as none of the runners installs pillow. Also all other tests that use this functionality are not run. The tests run fine on my machine but it would be nice to add testing on Travis as well. I could add installation of pillow to one/all Travis runners to also execute the tests here. Should this be done or is this unnecessary?

Generally, there were good reasons for SciPy to drop these functions [as discussed in the corresponding PR](https://github.com/scipy/scipy/pull/7869) so one might think about following their advice and moving to imageio and maybe skimage in the long run.

Commit:
Added scipy.misc.pilutil functions.

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-b0dd78fe2c552bd86a8cdc32e8801e9295c76fc1)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jnothman reviewed (commented):
Maybe this belongs under externals? Otherwise LGTM

Commit:
Added pillow to Travis and Appveyor for pilutils

jotasi commented:
Thanks @jnothman for the quick reply. I moved it to externals. I also added Pillow to Travis runners 1 & 2 to verify that the tests are passing there as well. Should I keep this or should I kick it out again?

jnothman reviewed (commented):
I don't think we want a directory sklearn/externals/tests. Would it hurt to have a sklearn/externals/pilutil/ directory?

jotasi commented:
Ok. I moved it into its own folder and subpackage. Also I added `_imsave`, which is used in `test_lfw.py`. I think now all calls to deprecated `scipy.misc.pilutil` functions are updated. Does anything else need to be changed?

jnothman commented on sklearn/datasets/lfw.py:
```diff
@@ -31,8 +31,8 @@
 
 from .base import get_data_home, _fetch_remote, RemoteFileMetadata
 from ..utils import Bunch
+from ..externals.pilutil import _imread, _imresize
```
How fast is importing pillow/pil on your machine? I would keep the import local if it will slow accessing datasets significantly just because the user happens to have the library installed

jotasi replied:
On my machine the import `from PIL import Image` takes approx. 0.04 seconds. The import `from sklearn.externals.pilutil import _imread, _imresize` takes 0.8 seconds though. Should I keep the global import or should I change it back to a local one?

jnothman reviewed (commented):
Otherwise lgtm

glemaitre commented on build_tools/travis/install.sh:
```diff
@@ -82,6 +82,10 @@ if [[ "$COVERAGE" == "true" ]]; then
     pip install coverage codecov
 fi
 
+if [[ "$PILLOW" == "true" ]]; then
```
Could you merge master because your install.sh is outdated: https://github.com/scikit-learn/scikit-learn/blob/master/build_tools/travis/install.sh#L40

I would probably add it to the `TO_INSTALL` environment variable if possible

jotasi replied:
I rebased on the current master and added the environment variable to `TO_INSTALL` for `"$DISTRIB" == "conda"`. The disadvantage is that I had to separately add it for the `"$DISTRIB" == "ubuntu"` case. Should I keep it like this?

glemaitre commented:
‎Would it be a good idea to skip the test for ubuntu and check that an error is raised and use the conda build for the functionality test. @lesteve WDYT? 

jotasi commented:
Atm. the tests are just skipped if pillow is not installed and thus no error would be raised. But I definitely can just add the test to one or both conda builds and skip them on the ubuntu one. My original reasoning for adding it to ubuntu was to check it on both python 2 and 3 (both conda builds are python 3, aren't they?).

glemaitre commented:
I was counting on appveyor for python 2. The only thing about this ubuntu build is that the version of numpy and scipy are older. 

jotasi commented:
Ok. I changed the last commit accordingly.

glemaitre commented:
yes you can. We were testing this case before (with scipy). I would also do it in circle CI.

jotasi commented:
I added pillow to appveyor and thus the tests in `test_pilutils.py` are now executed. In the circleci builds pillow was already installed before, but if I understood this correctly, these builds do not execute any tests, do they?

glemaitre commented:
Circle CI is just for the documentation. But we might need pillow to execute the example using the dataset which require pillow from now on.

> I added pillow to appveyor and thus the tests in test_pilutils.py are now executed.

Now that you mentioned, we don't need `test_pilutils.py`. We are not testing the package in externals usually. When I mentioned testing, it was about the internal tests which are using the `datasets` and therefore indirectly `imread` and `imresize`.

glemaitre commented on sklearn/externals/setup.py:
```diff
@@ -1,9 +1,13 @@
     config = Configuration('externals', parent_package, top_path)
     config.add_subpackage('joblib')
+    config.add_subpackage('pilutil')
+    config.add_subpackage('pilutil/tests')
```
So we can remove this line

glemaitre commented on sklearn/externals/setup.py:
```diff
@@ -1,9 +1,13 @@
     config.add_subpackage('joblib')
+    config.add_subpackage('pilutil')
+    config.add_subpackage('pilutil/tests')
+    config.add_data_dir(os.path.join('pilutil', 'tests', 'data'))
```
So we can remove this line

glemaitre commented on sklearn/externals/setup.py:
```diff
@@ -1,9 +1,13 @@
     from numpy.distutils.misc_util import Configuration
     config = Configuration('externals', parent_package, top_path)
     config.add_subpackage('joblib')
+    config.add_subpackage('pilutil')
```
we can move `_pilutil.py` to externals and name it `pilutils.py` and add `__all__` at the beginning of this file

glemaitre commented on sklearn/externals/setup.py:
```diff
@@ -1,9 +1,13 @@
 # -*- coding: utf-8 -*-
+import os
```
So we can remove this line

glemaitre reviewed (commented):
If we remove the tests, we don't need the data as well.
Otherwise LGTM

jotasi commented:
Thanks @glemaitre for the clarifications. I wasn't aware that generally no tests are included for modules in `sklearn.externals`. I moved `_pilutil.py` in externals and renamed it `pilutils.py` as you suggested. I also dropped the commit containing the tests. Is there anything else that should be changed?

glemaitre commented:
Only a single thing: add an entry in what's new. Otherwise LGTM on my side.

Please add an entry to the change log at doc/whats_new/v0.20.rst under bug fixes. Like the other entries there, please reference this pull request with :issue: and credit yourself (and other contributors if applicable) with :user:

jotasi commented:
Oh yes, I forgot that. An entry is added now. Thanks again.

glemaitre commented:
LGTM apart that I am not sure if Bug Fixes is the right section. @jnothman could you confirm that.

jotasi commented:
Ok I dropped the entry in whats_new again. Concerning your question about the duration of a global import, @jnothman, I had answered that directly under your comment (see comment to `sklearn/datasets/lfw.py `):

> On my machine the import from PIL import Image takes approx. 0.04 seconds. The import from sklearn.externals.pilutil import _imread, _imresize takes 0.8 seconds though. Should I keep the global import or should I change it back to a local one?

Testing it again on a different machine, the duration of the import `from sklearn.externals.pilutil import _imread, _imresize` reduces to 0.2 seconds. For my feeling this is not very costly and thus I kept the global import, but if you think this is too much of a delay, I can also change it to a local import again.

jnothman commented:
0.2s for rarely used functionality is quite a lot IMO. I'd make it local...

lesteve reviewed (commented):
Sorry a bit late to the party, but I am wondering whether pilutils.py belongs in externals. In my mind, externals is for packages that we vendor for convenience reasons and that we modify very little if at all. AFAICT sklearn/externals/pilutils.py is taken from scipy/misc/pilutil.py with some modifications, mostly removal of DeprecationWarning and renaming of functions.

Other than this I have a few comments.

jnothman commented:
i just think it's too big for utils and we don't care to test/lint it

jotasi commented on .travis.yml:
```diff
@@ -36,14 +36,14 @@ matrix:
     # Python 3.4 build
     - env: DISTRIB="conda" PYTHON_VERSION="3.4" INSTALL_MKL="false"
            NUMPY_VERSION="1.10.4" SCIPY_VERSION="0.16.1" CYTHON_VERSION="0.25.2"
-           COVERAGE=true
```
I changed it in both conda builds.

jotasi commented:
Thanks to all of you for your comments. I hope, I have addressed them all.

Concerning your comment @jnothman : I have made the import in sklearn/datasets/lfw.py and sklearn/datasets/base.py but not in the tests. Does this cover your concern?

Concerning your comments @lesteve : The format of the package addition to the Travis builds was changed to match the other packages. I also changed the filename to pilutil.py and added a reference to the original file [at the top of pilutil.py](https://github.com/jotasi/scikit-learn/blob/2ad69e406df3b01bbee5338c0378900756e6b9b4/sklearn/externals/pilutil.py#L7-L8). I have not changed the function names yet. My reasoning for the underscores was that these functions actually are only used internally and are not really supposed to be part of the public API. That is also why I have removed all parameters that are not used in sklearn as I thought this might actually be clearer and easier to maintain. Therefore, I felt that using the same names as in the original would be misleading. Do you prefer to keep the (longer and more complicated) scipy-version more or less as is?

jnothman commented:
> Concerning your comment @jnothman : I have made the import in sklearn/datasets/lfw.py and sklearn/datasets/base.py but not in the tests. Does this cover your concern?

Looks good.

Btw, please avoid amending your commits. It's much easier for us to review changes if you simply add new commits.

LGTM

jotasi commented:
Thanks for the info. So far I mostly worked on projects, where contributors where asked to (at least before merging) clean up their commit history. I'll refrain from changing my history here, then.

Also again, thanks to all of you for all your help!

jnothman commented:
we used to do that, but GitHub made it easy to squash commits when merging,
and provides benefits for reviewing when commits are only added

lesteve commented:
> I have not changed the function names yet. My reasoning for the underscores was that these functions actually are only used internally and are not really supposed to be part of the public API. That is also why I have removed all parameters that are not used in sklearn as I thought this might actually be clearer and easier to maintain. Therefore, I felt that using the same names as in the original would be misleading. Do you prefer to keep the (longer and more complicated) scipy-version more or less as is?

This is only my opinion and happy to hear if others disagree. I am fine with keeping pilutil.py in externals, but I would be in favour of a unmodified copy of scipy/misc/pilutil.py. It's probably fine to remove some functions if they are not used but I'd like to keep the diff as simple as possible between the scipy pilutil.py and our pilutil.py. I would not bother with adding underscores.

To be perfectly honest, I am not sure where we are going long-term with this approach. Basically we are telling scipy: you are deprecating some functions and we want to keep them, so we are going to copy and paste your code and maintain it for some time (maybe until there is a clearer picture of what we can use as a replacement?). Rather than copying and pasting the scipy code, why not just catch the DeprecationWarning and define `imread` et al in `sklearn/util/fixes.py` along these lines? 

```py
from sklearn.utils.testing import ignore_warnings
def imread(*args, **kwargs):
    from scipy.misc import imread  # need some ImportError handling here
    with warnings.catch_warnings(category=DeprecationWarning):
        imread(*args, **kwargs)
```

I feel this is a more minimal code change and buy us time for 2 scipy releases (1-2 year at the current pace of scipy release). Then we can decide whether copying and pasting the pilutil.py code is the best option.

Again, more than happy to hear other opinions on this ...

jnothman commented:
I suppose we could just catch and hide the DeprecationWarning. I can't say
it feels particularly comfortable. It means, though, that actually need to
make a decision quite soon (lest we forget about it and it all breaks),
thus not buying us much time at all.

On 16 January 2018 at 21:15, Loïc Estève <notifications@github.com> wrote:

> I have not changed the function names yet. My reasoning for the
> underscores was that these functions actually are only used internally and
> are not really supposed to be part of the public API. That is also why I
> have removed all parameters that are not used in sklearn as I thought this
> might actually be clearer and easier to maintain. Therefore, I felt that
> using the same names as in the original would be misleading. Do you prefer
> to keep the (longer and more complicated) scipy-version more or less as is?
>
> This is only my opinion and happy to hear if others disagree. I am fine
> with keeping pilutil.py in externals, but I would be in favour of a
> unmodified copy of scipy/misc/pilutil.py. It's probably fine to remove some
> functions if they are not used but I'd like to keep the diff as simple as
> possible between the scipy pilutil.py and our pilutil.py. I would not
> bother with adding underscores.
>
> To be perfectly honest, I am not sure where we are going long-term with
> this approach. Basically we are telling scipy: you are deprecating some
> functions and we want to keep them, so we are going to copy and paste your
> code and maintain it for some time (maybe until there is a clearer picture
> of what we can use as a replacement?). Rather than copying and pasting the
> scipy code, why not just catch the DeprecationWarning and define imread
> et al in sklearn/util/fixes.py along these lines?
>
> from sklearn.utils.testing import ignore_warningsdef imread(*args, **kwargs):
>     from scipy.misc import imread  # need some ImportError handling here
>     with warnings.catch_warnings(category=DeprecationWarning):
>         imread(*args, **kwargs)
>
> I feel this is a more minimal code change and buy us time for 2 scipy
> releases (1-2 year at the current pace of scipy release). Then we can
> decide whether copying and pasting the pilutil.py code is the best option.
>
> Again, more than happy to hear other opinions on this ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10427#issuecomment-357914639>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69i0vxuBNbtiLXO_HUViYhQTcCGpks5tLHbTgaJpZM4RWz7J>
> .
>

glemaitre commented:
Could we settle down on a fix directly? I see only 2 solutions:• write a utils/image.py with the utils we need and the tests. We can probably use the scipy code. We will need PIL as dependency as now. • required ‎skimage as a optional dependency. I would rather think to go for the first point. Delaying the issue will not solved it, IMO.

lesteve commented:
A third option is the approach taken in this PR (at the time of writing): have sklearn/externals/pilutil.py be a copy of scipy/misc/pilutil.py with minimal changes (personally I would say removal of DeprecationWarning and removal of functions we do not need but no more than that). We can even call the file `_pilutil.py` if we want to indicate that users should not import `sklearn.externals._pilutil` directly.

If we are planning on maintaining a copy of the scipy code for the foreseeable future, I feel like option 1. is the right thing to do. If we are fine taking a copy of the scipy code, putting it in externals and think that there is a very small likelihood that we have to change it, option 3. is fine and probably less work.

glemaitre commented:
> If we are fine taking a copy of the scipy code, putting it in externals and think that there is a very small likelihood that we have to change it, option 3. is fine and probably less work.

Then, I would go for option 3 in this case, and let the file pilutils as close as possible to the original file.

glemaitre commented:
FWIIW, `pilutils.py` had very little changes for the part that we need (documentation mainly) and I don't think that we will have to touch it.

Commit:
Added back unused parts to reduce diff.

jotasi commented:
As option 3 seems to be the consensus, I've updated the PR: `sklearn.externals._pilutil` now contains the original versions of all functions used from `scipy.misc.pilutil`. The only things that were changed are:

 - When pillow is not installed, the module does not fail at import but when a function is called, that requires pillow.
 - The documentation was changed slightly and the license was added.
 - The example for `imsave` was removed as it would fail on machines without pillow and did not seem to be overly important.

PS: The Travis build is canceled before running not just here, but for all PRs that were updated before. There seems to be something wrong with it. (On my fork, Travis is running fine...)

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-31783cd359aff99d4cd5d1a5f274198515b17cf7)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

lesteve committed:
Tweaks to _pilutil.py to reduce differences from original scipy pilutil.py

lesteve committed:
Tweak comments

lesteve commented:
I pushed some tweaks , mostly to reduce differences with respect to the original scipy file. I think I am fine with this as it is.

Probably best to leave it for another PR, but there seems to be some occurences of imread/imresize in doc and examples:
```
doc/datasets/index.rst:* `scipy.misc.imread <https://docs.scipy.org/doc/scipy/reference/generated/scipy.
doc/datasets/index.rst:  misc.imread.html#scipy.misc.imread>`_ (requires the `Pillow
examples/classification/plot_digits_classification.py:# matplotlib.pyplot.imread.  Note that each image must have the same size. For 
examples/cluster/plot_face_segmentation.py:face = sp.misc.imresize(face, 0.10) / 255.
examples/cluster/plot_face_ward_segmentation.py:face = sp.misc.imresize(face, 0.10) / 255.
```

lesteve committed:
Minor tweaks

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-42c56625cd1bd108a20ab55e71c10d3fc2f39a78)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

lesteve committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into pr/10427

lesteve commented:
I pulled master to get https://github.com/scikit-learn/scikit-learn/pull/10486 which should (hopefully) make Travis happy.

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-35a490acdb3ded66c1232633ca8306149794ca3d)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jnothman commented:
I agree with lgtm.com at first glance! It looks like there is a valid code
path where strdata is unset!

On 18 January 2018 at 01:59, sklearn-lgtm <notifications@github.com> wrote:

> This pull request introduces 1 alert - view on lgtm.com
> <https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-35a490acdb3ded66c1232633ca8306149794ca3d>
>
> *new alerts:*
>
>    - 1 for Potentially uninitialized local variable
>
> ------------------------------
>
> *Comment posted by lgtm.com <https://lgtm.com>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10427#issuecomment-358330259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63T9oT7R9ke05L_nzwx0j7fpi6seks5tLgq2gaJpZM4RWz7J>
> .
>

jotasi commented:
First of all: Thanks @lesteve for cleaning up after me!

Concerning the complaint of lgtm.com:
As far as I can see this cannot happen, although, if there were no restriction to keep the diff small I would add some input validation.
At the beginning of `toimage` `shape` is checked to be of either length 2 or 3:
https://github.com/jotasi/scikit-learn/blob/2f78ea7653515be83b837399310bc9621a3feb31/sklearn/externals/_pilutil.py#L357-L361
Therefore, this line, taking the `ca`th element of `shape` would fail for any `ca` not in [0,1,2]:
https://github.com/jotasi/scikit-learn/blob/2f78ea7653515be83b837399310bc9621a3feb31/sklearn/externals/_pilutil.py#L410
And thus this block will never be called with `ca` not in [0,1,2]:
https://github.com/jotasi/scikit-learn/blob/2f78ea7653515be83b837399310bc9621a3feb31/sklearn/externals/_pilutil.py#L415-L423
Although, I do agree that the error message, that the list index of `shape` is out of range, is not a very informative error message.

Concerning the errors on Travis:
Sadly, the pull of the master did not fix the problem. I think for the ignore option to work one needs to specify the full (absolute or relative) path to the folder/file one wants to ignore, rather than the path from the modules root. At least that is the case on my local machine.

lesteve commented:
I would ignore the lgtm warning, because that's in externals, for the same reason that we do not test code in externals.

About the error in Travis, looks like the `--ignore` does not work because we are using the installed sklearn package through `--pyargs` (and not the local folder). I'll try to figure out a way of doing it. If there is no easy way, removing the doctests in `_pilutil.py` is reasonable.

lesteve commented:
I just pushed an alternative solution for not running tests in externals, fingers crossed.

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-af85dd275da0ae21d6fb154b0e3a86cfa4d0ffa6)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

lesteve committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into pr/10427

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-6334984935d3575750fb6c78ad339c2e2fc6166e)

**new alerts:**

* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

lesteve commented:
LGTM, merging, thanks a lot @jotasi!

amueller commented:
Sorry I'm out of the loop, but wouldn't it be better to add a dependency to one of the libraries? That will point people more directly towards the right solution. And we only need this in some rare cases.

jnothman commented:
What do you mean by adding a dependency to one of the libraries?

glemaitre commented:
> Sorry I'm out of the loop, but wouldn't it be better to add a dependency to one of the libraries? That will point people more directly towards the right solution. And we only need this in some rare cases.

To be able to read image, we will need at least pillow + numpy and make the conversion ourself. This is what imageio is doing in fact. Resizing is a more complex thing since that we would need to use scikit-image which would required dependencies as well.
